libs/elasticsearch/langchain_elasticsearch/cache.py

```diff
@@ -186,3 +236,213 @@ def clear(self, **kwargs: Any) -> None:
+                    source_includes=["vector_dump", "timestamp"],
+                )
+
+            except BadRequestError as e:
+                if "window too large" in (
+                    e.body.get("error", {}).get("root_cause", [{}])[0].get("reason", "")
+                ):
+                    logger.warning(
+                        "Exceeded the maximum window size, "
+                        "Reduce the duplicates manually or lower "
+                        "`maximum_duplicate_allowed.`"
+                    )
+                    raise e
+
+            total_hits = results["hits"]["total"]["value"]
+            if self._maximum_duplicate_allowed > 1 and total_hits > len(cache_keys):
+                logger.warning(
+                    f"Deduplicating, found {total_hits} hits for {len(cache_keys)} keys"
+                )
+                map_ids = self._deduplicate_hits(results["hits"]["hits"])
+            else:
+                map_ids = {
+                    r["_id"]: r["_source"]["vector_dump"]
+                    for r in results["hits"]["hits"]
+                }
+
+            return [map_ids.get(k) for k in cache_keys]
+
+        else:
+            records = self._es_client.mget(
+                index=self._index_name, ids=cache_keys, source_includes=["vector_dump"]
+            )
+            return [
+                r["_source"]["vector_dump"] if r["found"] else None
+                for r in records["docs"]
+            ]
+
+    def build_document(self, text_input: str, vector: List[float]) -> Dict[str, Any]:
+        """Build the Elasticsearch document for storing a single embedding"""
+        body: Dict[str, Any] = {
+            "vector_dump": vector,
+            "timestamp": datetime.now().isoformat(),
+        }
+        if self._metadata is not None:
+            body["metadata"] = self._metadata
+        if self._store_input:
+            body["text_input"] = text_input
+        if self._namespace:
+            body["namespace"] = self._namespace
+        return body
+
+    def _bulk(self, actions: Iterable[Dict[str, Any]]) -> None:
+        try:
+            helpers.bulk(
+                client=self._es_client,
+                actions=actions,
+                index=self._index_name,
+                require_alias=self._is_alias,
+                refresh=True,
+            )
+        except BulkIndexError as e:
+            first_error = e.errors[0].get("index", {}).get("error", {})
+            logger.error(f"First bulk error reason: {first_error.get('reason')}")
+            raise e
+
+    def mset(self, key_value_pairs: Sequence[Tuple[str, List[float]]]) -> None:
+        """Set the values for the given keys."""
+        actions = (
+            {
+                "_op_type": "index",
+                "_id": self._key(key),
+                "_source": self.build_document(key, vector),
+            }
+            for key, vector in key_value_pairs
+        )
+        self._bulk(actions)
+        return
```

Comment:
```suggestion
```
I believe we don't need to have this explicitly

Reply:
done